Spring data creteria query hw solution

JohnGrey17 opened the pull request.

JohnGrey17 committed:
removed BookRepository implementation Class instead that did extend JpaRepositoty Interface

JohnGrey17 committed:
changer parameter im spring.jpa.hibernate.ddl-auto on -validate

JohnGrey17 committed:
added changeLog configuration

JohnGrey17 committed:
Implemented PUT request

JohnGrey17 committed:
implemented soft delete (added isdeleted to Entity and yaml file)
implemented method on service layer and controller

JohnGrey17 committed:
implemented soft delete (added isdeleted to Entity and yaml file)
implemented method on service layer and controller

JohnGrey17 committed:
implemented soft delete (added isdeleted to Entity and yaml file)
implemented method on service layer and controller

JohnGrey17 committed:
fixed update method in service layer
changed deprecated @WHERE annotation on @SQLRestriction

JohnGrey17 committed:
MVN clean package

JohnGrey17 committed:
added environment variable dor password and login

JohnGrey17 committed:
fixed implementation for updateBookById in service layer

JohnGrey17 committed:
fixed solution

JohnGrey17 committed:
Added BookSearchParameter class to DTO folder
added specificationBuilderInterface
added bookSpecificationBuilderImpl(without implementation)

JohnGrey17 committed:
Added BookSearchParameter class to DTO folder
added specificationBuilderInterface
added bookSpecificationBuilderImpl(without implementation)

JohnGrey17 committed:
implemented search method

JohnGrey17 committed:
implemented search method

JohnGrey17 committed:
fixed search method

JohnGrey17 committed:
fixed solution

andrii-hoienko commented on src/main/java/org/example/bookstore/repository/book/BookSpecificationBuilder.java:
```diff
@@ -0,0 +1,33 @@
+    public Specification<Book> build(BookSearchParameters searchParameters) {
+        Specification<Book> spec = Specification.where(null);
+
+        if (searchParameters.title() != null && searchParameters.title().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager
+                    .getSpecificationProvider("title")
+                    .getSpecification(searchParameters.title()));
+        }
+        if (searchParameters.author() != null && searchParameters.author().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager
+                    .getSpecificationProvider("author")
+                    .getSpecification(searchParameters.author()));
+        }
+        return spec;
```
I see at least 2 code duplications:

1. Check if the string is null or empty. I suggest to create some util class for such checks or use already existing libraries like Apache common utils
2. Code that populates specification is common for both author and title, so I suggest extracting it in method and reuse. This will make code more flexible because it will be much easier to add new search parameters, maybe event strategy will be suitable here but it may be overcomplication, so just make a method for now

andrii-hoienko commented on src/main/java/org/example/bookstore/repository/book/BookSpecificationBuilder.java:
```diff
@@ -0,0 +1,33 @@
+
+        if (searchParameters.title() != null && searchParameters.title().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager
+                    .getSpecificationProvider("title")
+                    .getSpecification(searchParameters.title()));
+        }
+        if (searchParameters.author() != null && searchParameters.author().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager
+                    .getSpecificationProvider("author")
+                    .getSpecification(searchParameters.author()));
+        }
```
Strings duplication, please make them constants

andrii-hoienko reviewed (changes requested):
Good job, but some improvements required

JohnGrey17 committed:
added addSpecification method

karma-o commented on src/main/java/org/example/bookstore/repository/book/spec/TitleSpecificationProvider.java:
```diff
@@ -0,0 +1,22 @@
+
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    private static final String TITLE = "title";
```
I would reuse this constant in the builder class as well instead of creating a duplicate, same for author

karma-o reviewed (approved):
Nice!

JohnGrey17 committed:
Add functionality pagination sorting swagger (#7)

* Implemented Validator for DTO and CustomGlobalExceptionHandler with implementation for validator and EntityNotFoundException

implemented pageable interface in all layers
added Swager and description of the method in the control layer


* deleted InvalidPriceException
deleted annotation @Digit for BigInteger variable
added custom validator for ISBN

* Added @Tag annotation
removed @Digit annotation

* Added @Tag annotation
added annotation for ISBN
added Pageable variable to the search method

* Added @Operation for the search method
Added check for delete method
